p2p/pex/addrbook.go

```diff
@@ -324,10 +329,24 @@ func (a *addrBook) MarkAttempt(addr *p2p.NetAddress) {
 	ka.markAttempt()
 }
 
-// MarkBad implements AddrBook. Currently it just ejects the address.
-// TODO: black list for some amount of time
+// MarkBad implements AddrBook. Kicks address out from book, places
+// the address in the badPeers pool.
 func (a *addrBook) MarkBad(addr *p2p.NetAddress) {
-	a.RemoveAddress(addr)
+	if a.addBadPeer(addr) {
+		a.RemoveAddress(addr)
```

Comment:
`addBadPeer()` returns `false` if the peer is already considered bad, in which case it will not be removed from the address book. Shouldn't we always remove it regardless, since we do not want bad peers there under any circumstances? We may also want to refresh the ban timer in that case.

Also, since both of these calls take out mutex locks separately, the data can change between them, which may cause problems. Not sure if that will be an actual problem here, but in general it's a good idea to take out mutex locks in public methods, and have internal methods that assume locks are already held such that they can be combined and composed.

Reply:
I guess if it's bad but still in part of the address book then we should return true so that it still runs RemoveAddress. 

For your second paragraph about the mutexes, that makes sense

Reply:
actually I remember why I didn't add the mutexes in the public function because RemoveAddress has it's own mutex

Reply:
Exactly. Which means that `addBadPeer` and `RemoveAddress` can't be called in a single critical section. If we had `removeAddress()` which did not take out a lock, and `RemoveAddress()` which does take out a lock and call `removeAddress()` internally, then we could use `removeAddress()` in a critical section together with `addBadPeer()`.

Again, not sure if it's worth spending time on in this case, since I suspect it's safe to call `RemoveAddress()` regardless of any state changes. But this is a general problem across the code base which we should keep in mind.

Reply:
ah yes I understand. I think I will change it like that anyway